### PR TITLE
expression: rewrite builtin function: FIELD

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

File: expression/builtin_string_test.go [hunk 1]
```diff
@@ -1214,7 +1220,12 @@ func (s *testEvaluatorSuite) TestField(c *C) {
 		fc := funcs[ast.Field]
 		f, err := fc.getFunction(s.ctx, datumsToConstants(types.MakeDatums(t.argLst...)))
 		c.Assert(err, IsNil)
+		c.Assert(f, NotNil)
+		c.Assert(f.canBeFolded(), IsTrue)
 		r, err := f.eval(nil)
+		if err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: remove these

File: expression/integration_test.go [hunk 2]
```diff
@@ -834,6 +834,16 @@ func (s *testIntegrationSuite) TestStringBuiltin(c *C) {
...
 	result.Check(testkit.Rows("'121' '0' '中文' <nil>"))
+
+	// for field
+	result = tk.MustQuery(`select field(1, 2, 1), field(1, 0, NULL), field(1, NULL, 2, 1), field(NULL, 1, 2, NULL);`)
+	result.Check(testkit.Rows("2 0 3 0"))
+	result = tk.MustQuery(`select field("1", 2, 1), field(1, "0", NULL), field("1", NULL, 2, 1), field(NULL, 1, "2", NULL);`)
+	result.Check(testkit.Rows("2 0 3 0"))
+	result = tk.MustQuery(`select field("1", 2, 1), field(1, "abc", NULL), field("1", NULL, 2, 1), field(NULL, 1, "2", NULL);`)
+	result.Check(testkit.Rows("2 0 3 0"))
+	result = tk.MustQuery(`select field("abc", "a", 1);`)
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: add test for select(1.3, "1.3", ...)

File: expression/builtin_string.go [hunk 3]
```diff
@@ -1960,61 +1963,105 @@ func (c *fieldFunctionClass) getFunction(ctx context.Context, args []Expression)
...
-// Returns 0 if arg0 is not found.
-// If arg0 is NULL, the return value is 0 because NULL fails equality comparison with any value.
-// If all arguments are strings, all arguments are compared as strings.
-// If all arguments are numbers, they are compared as numbers.
-// Otherwise, the arguments are compared as double.
-func (b *builtinFieldSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+func (b *builtinFieldIntSig) evalInt(row []types.Datum) (int64, bool, error) {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: add comment here

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by zz-jason** [comment]
@jackysp @breeswish PTAL

**Pull request merged**